zipline/algorithm.py

```diff
@@ -407,7 +407,8 @@ def before_trading_start(self, data):
         self._in_before_trading_start = False
 
     def handle_data(self, data):
-        self._handle_data(self, data)
+        if self._handle_data:
```

Comment:
I would have preferred to check if `self._handle_data` was defined at the time of scheduling the `Always()` call to `handle_data`, but we have various tests that attach a `handle_data` after the `TradingAlgorithm` is created.  Easier to leave it this way for now.